pkg/scheduler/plugins/topology/topology_plugin_job_filtering.go

```diff
@@ -0,0 +1,277 @@
+	var jobAllocateableDomain []*TopologyDomainInfo
+	if maxAllocatablePods >= len(podgroup_info.GetTasksToAllocate(job, t.taskOrderFunc, true)) {
+		jobAllocateableDomain, err = t.getBestjobAllocateableDomains(job, topologyTree)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Clean allocation data from the tree
+	for _, levelDomains := range topologyTree.DomainsByLevel {
+		for _, domain := range levelDomains {
+			domain.AllocatablePods = 0
+		}
+	}
+
+	if len(jobAllocateableDomain) == 0 {
+		log.InfraLogger.V(6).Infof("no relevant domains found for job %s, workload topology name: %s",
+			job.PodGroup.Name, topologyTree.Name)
+	}
+
+	return nil
+}
+
+func (t *topologyPlugin) getJobTopology(job *podgroup_info.PodGroupInfo) (*TopologyInfo, error) {
+	jobTopologyName := job.PodGroup.Spec.TopologyConstraint.Topology
+	if jobTopologyName == "" {
+		return nil, nil
+	}
+	topologyTree := t.TopologyTrees[jobTopologyName]
+	if topologyTree == nil {
+		return nil, fmt.Errorf("matching topology tree haven't been found for job %s, workload topology name: %s",
+			job.PodGroup.Name, jobTopologyName)
+	}
+	return topologyTree, nil
+}
+
+func (t *topologyPlugin) calcTreeAllocatable(job *podgroup_info.PodGroupInfo, topologyTree *TopologyInfo) (int, error) {
+	jobAllocationMetaData, err := initJobAllocationMetadataStruct(job, t)
+	if err != nil {
+		return 0, err
+	}
+
+	return t.calcSubTreeAllocatable(jobAllocationMetaData, topologyTree.Root)
+}
+
+func initJobAllocationMetadataStruct(job *podgroup_info.PodGroupInfo, t *topologyPlugin) (*jobAllocationMetaData, error) {
+	tasksToAllocate := podgroup_info.GetTasksToAllocate(job, t.taskOrderFunc, true)
+	maxPodResources := resource_info.NewResourceRequirements(0, 0, 0)
+	for _, podInfo := range tasksToAllocate {
+		err := maxPodResources.SetMaxResource(podInfo.ResReq)
+		if err != nil {
+			return nil, err
+		}
+	}
+	initialAllocationTestPods := []*pod_info.PodInfo{
+		{Name: "1-pods-resources", ResReq: maxPodResources},
+	}
+	jobAllocationData := &jobAllocationMetaData{
+		maxPodResources:    maxPodResources,
+		allocationTestPods: initialAllocationTestPods,
+		tasksToAllocate:    tasksToAllocate,
+	}
+	return jobAllocationData, nil
+}
+
+func (t *topologyPlugin) calcSubTreeAllocatable(jobAllocationData *jobAllocationMetaData, rootDomain *TopologyDomainInfo) (int, error) {
+	if rootDomain == nil {
+		return 0, nil
+	}
+
+	if len(rootDomain.Children) == 0 {
+		for _, node := range rootDomain.Nodes {
+			rootDomain.AllocatablePods += calcNodeAccomedation(jobAllocationData, node)
+		}
+		return rootDomain.AllocatablePods, nil
+	}
+
+	for _, child := range rootDomain.Children {
+		childAllocateable, err := t.calcSubTreeAllocatable(jobAllocationData, child)
+		if err != nil {
+			return 0, err
+		}
+		rootDomain.AllocatablePods += childAllocateable
+	}
+	return rootDomain.AllocatablePods, nil
+}
+
+func calcNodeAccomedation(jobAllocationMetaData *jobAllocationMetaData, node *node_info.NodeInfo) int {
```

Comment:
Please check PRs for typos, it looks bad and I can't comment on them individually. Install a spell checker extension it's super easy